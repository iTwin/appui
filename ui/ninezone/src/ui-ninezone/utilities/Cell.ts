/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

/** Describes [[Cell]]. */
export interface CellProps {
  row: number;
  col: number;
}

/** Provides methods to work with cells. */
export class Cell implements CellProps {
  /** Creates cell from [[CellProps]]. */
  public static create(props: CellProps) {
    return new Cell(props.row, props.col);
  }

  private _row: number;
  private _col: number;

  /** @returns Row of this cell. */
  public get row() {
    return this._row;
  }

  /** @returns Column of this cell. */
  public get col() {
    return this._col;
  }

  /** Creates a new cell. */
  public constructor(row: number, col: number) {
    this._row = row;
    this._col = col;
  }

  /** @returns True if this and other cells are equal.  */
  public equals(other: CellProps) {
    return this.row === other.row && this.col === other.col;
  }

  /** @returns True if this and other cells are on same row, but different columns. */
  public isRowAlignedWith(other: CellProps) {
    if (this.row === other.row)
      if (this.col !== other.col)
        return true;
    return false;
  }

  /** @returns True if this and other cells are on same column, but different rows. */
  public isColumnAlignedWith(other: CellProps) {
    if (this.col === other.col)
      if (this.row !== other.row)
        return true;
    return false;
  }

  /** @returns Column aligned cells between this and other cells. */
  public getVerticallyAlignedCellsTo(other: CellProps) {
    const cells: CellProps[] = [];

    if (!this.isColumnAlignedWith(other))
      return cells;

    const diff = this.row > other.row ? -1 : 1;
    for (let i = this.row + diff; i !== other.row; i += diff) {
      cells.push(new Cell(i, this.col));
    }

    return cells;
  }

  /** @returns Row aligned cells between this and other cells. */
  public getHorizontallyAlignedCellsTo(other: CellProps) {
    const cells: CellProps[] = [];

    if (!this.isRowAlignedWith(other))
      return cells;

    const diff = this.col > other.col ? -1 : 1;
    for (let i = this.col + diff; i !== other.col; i += diff) {
      cells.push(new Cell(this.row, i));
    }

    return cells;
  }

  /** @returns Row or column aligned cells between this and other cells. */
  public getAlignedCellsTo(other: CellProps) {
    const cells = this.getHorizontallyAlignedCellsTo(other);
    if (cells.length > 0)
      return cells;

    return this.getVerticallyAlignedCellsTo(other);
  }

  /** @returns True if this cell is between cell1 and cell2 on vertical axis. */
  public isVerticallyBetween(cell1: CellProps, cell2: CellProps) {
    if (this.isColumnAlignedWith(cell1) && this.isColumnAlignedWith(cell2))
      if ((this.row < cell1.row && this.row > cell2.row) || this.row > cell1.row && this.row < cell2.row)
        return true;

    return false;
  }

  /** @returns True if this cell is between cell1 and cell2 on horizontal axis. */
  public isHorizontallyBetween(cell1: CellProps, cell2: CellProps) {
    if (this.isRowAlignedWith(cell1) && this.isRowAlignedWith(cell2))
      if ((this.col < cell1.col && this.col > cell2.col) || this.col > cell1.col && this.col < cell2.col)
        return true;

    return false;
  }

  /** @returns True if this cell is between cell1 and cell2 (column aligned or row aligned).  */
  public isBetween(cell1: CellProps, cell2: CellProps) {
    if (this.isVerticallyBetween(cell1, cell2))
      return true;
    return this.isHorizontallyBetween(cell1, cell2);
  }
}
