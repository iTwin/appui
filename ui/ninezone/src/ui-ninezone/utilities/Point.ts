/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

/** Describes [[Point]].
 * @beta
 */
export interface PointProps {
  readonly x: number;
  readonly y: number;
}

/** Describes and provides methods to work with 2d points.
 * @alpha
 */
export class Point implements PointProps {
  /** Creates point from [[PointProps]]. */
  public static create(pointProps: PointProps) {
    return new Point(pointProps.x, pointProps.y);
  }

  /** Creates a new point. */
  public constructor(public readonly x = 0, public readonly y = 0) {
  }

  /** Calculates distance to other point. */
  public getDistanceTo(other: PointProps): number {
    const offset = this.getOffsetTo(other);
    return Math.sqrt(Math.pow(offset.x, 2) + Math.pow(offset.y, 2));
  }

  /** Gets offset to other point. */
  public getOffsetTo(other: PointProps) {
    return new Point(other.x - this.x, other.y - this.y);
  }

  /** @returns New [[Point]] that is offset along the X and Y axes. */
  public offset(offset: PointProps) {
    return new Point(this.x + offset.x, this.y + offset.y);
  }

  /** @returns New [[Point]] that is offset along the X axis. */
  public offsetX(offset: number) {
    return new Point(this.x + offset, this.y);
  }

  /** @returns New [[Point]] that is offset along the Y axis. */
  public offsetY(offset: number) {
    return new Point(this.x, this.y + offset);
  }

  /** @returns True if position of this and other points are equal. */
  public equals(other: PointProps) {
    return other.x === this.x && other.y === this.y;
  }

  /** @returns New [[Point]] with modified x value. */
  public setX(x: number) {
    return new Point(x, this.y);
  }

  /** @returns New [[Point]] with modified y value. */
  public setY(y: number) {
    return new Point(this.x, y);
  }
}
