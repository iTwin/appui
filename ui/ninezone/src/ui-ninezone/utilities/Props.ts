/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

import * as React from "react";
import { Omit } from "@bentley/ui-core";

/** Props used by components that do not expect children to be passed in.
 * @alpha Move to ui-core
 */
export interface NoChildrenProps {
  children?: undefined;
}

/** Omit children property from T.
 * @internal
 */
export type OmitChildrenProp<T extends { children?: React.ReactNode; }> = Omit<T, "children">;

/** Flattens react fragments.
 * @internal
 */
// tslint:disable-next-line:variable-name
export const FlattenChildren = (children: React.ReactNode): React.ReactNode => {
  const items = React.Children.map(children, (child) => {
    if (!React.isValidElement<{ children?: React.ReactNode }>(child))
      return child;

    if (child.type === React.Fragment) {
      return child.props.children;
    }

    return child;
  });
  return items;
};
