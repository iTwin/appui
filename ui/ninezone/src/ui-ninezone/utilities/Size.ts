/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

/** Describes [[Size]].
 * @beta
 */
export interface SizeProps {
  readonly width: number;
  readonly height: number;
}

/** Describes 2d dimensions.
 * @beta
 */
export class Size implements SizeProps {
  /** Creates a size from [[SizeProps]]. */
  public static create(size: SizeProps) {
    return new Size(size.width, size.height);
  }

  /** Creates a size with specified dimensions. */
  public constructor(public readonly width = 0, public readonly height = 0) {
  }

  /** Checks if dimensions of two sizes are equal. */
  public equals(other: SizeProps) {
    if (this.width === other.width &&
      this.height === other.height)
      return true;
    return false;
  }
}
