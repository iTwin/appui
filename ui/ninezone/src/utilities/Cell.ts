/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Utilities */

/** Describes [[Cell]]. */
export interface CellProps {
  row: number;
  col: number;
}

/** Provides methods to work with cells. */
export default class Cell implements CellProps {
  /** Creates cell from [[CellProps]]. */
  public static create(props: CellProps) {
    return new Cell(props.row, props.col);
  }

  private _row: number;
  private _col: number;

  /** @returns Row of this cell. */
  public get row() {
    return this._row;
  }

  /** @returns Column of this cell. */
  public get col() {
    return this._col;
  }

  /** Creates a new cell. */
  public constructor(row: number, col: number) {
    this._row = row;
    this._col = col;
  }

  /** @returns True if this and other cells are equal.  */
  public equals(other: CellProps) {
    return this.row === other.row && this.col === other.col;
  }

  /** @returns True if this and other cells are on same row, but different columns. */
  public isRowAlignedWith(other: CellProps) {
    if (this.row === other.row)
      if (this.col !== other.col)
        return true;
    return false;
  }

  /** @returns True if this and other cells are on same column, but different rows. */
  public isColumnAlignedWith(other: CellProps) {
    if (this.col === other.col)
      if (this.row !== other.row)
        return true;
    return false;
  }

  /** @returns Row or column aligned cells between this and other cells. */
  public getAlignedCellsTo(other: CellProps) {
    const cells: CellProps[] = [];

    cells.push(new Cell(this.row, this.col));

    if (this.isRowAlignedWith(other)) {
      let diff = 1;
      if (this.col > other.col)
        diff = -1;

      let i = this.col;
      do {
        i += diff;
        cells.push(new Cell(this.row, i));
      } while (i !== other.col);
    } else if (this.isColumnAlignedWith(other)) {
      let diff = 1;
      if (this.row > other.row)
        diff = -1;

      let i = this.row;
      do {
        i += diff;
        cells.push(new Cell(i, this.col));
      } while (i !== other.row);
    }

    return cells;
  }

  /** @returns True if this cell is between cell1 and cell2 (column aligned or row aligned).  */
  public isBetween(cell1: CellProps, cell2: CellProps) {
    const c1 = Cell.create(cell1);
    if (c1.isRowAlignedWith(cell2) && this.row === c1.row) {
      if ((this.col < cell1.col && this.col > cell2.col) || this.col > cell1.col && this.col < cell2.col)
        return true;
    } else if (c1.isColumnAlignedWith(cell2) && this.col === c1.col)
      if ((this.row < cell1.row && this.row > cell2.row) || this.row > cell1.row && this.row < cell2.row)
        return true;

    return false;
  }
}

export class CellPropsHelpers {
  public static createCell(cellProps: CellProps): Cell {
    return new Cell(cellProps.row, cellProps.col);
  }
}
