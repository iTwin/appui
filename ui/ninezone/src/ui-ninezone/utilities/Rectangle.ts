/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

import { Point, PointProps } from "./Point";
import { Size, SizeProps } from "./Size";

/** Describes [[Rectangle]]. */
export interface RectangleProps {
  readonly left: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
}

/** Available corners of [[Rectangle]]. */
export enum Corner {
  TopLeft,
  TopRight,
  BottomRight,
  BottomLeft,
}

/** Available edges of [[Rectangle]]. */
export enum Edge {
  Left,
  Top,
  Right,
  Bottom,
}

/** Describes and provides methods to work with 2d bounds. */
export class Rectangle implements RectangleProps {
  /** Creates rectangle from [[RectangleProps]]. */
  public static create(props: RectangleProps) {
    return new Rectangle(props.left, props.top, props.right, props.bottom);
  }

  /** Creates rectangle from [[SizeProps]]. */
  public static createFromSize(size: SizeProps) {
    return new Rectangle(0, 0, size.width, size.height);
  }

  /** Creates rectangle with specified bounds. */
  public constructor(public readonly left = 0, public readonly top = 0, public readonly right = 0, public readonly bottom = 0) {
  }

  /** @returns Size of this rectangle. */
  public getSize() {
    const width = this.getWidth();
    const height = this.getHeight();
    return new Size(width, height);
  }

  /** @returns Width of this rectangle. */
  public getWidth() {
    return this.right - this.left;
  }

  /** @returns Height of this rectangle. */
  public getHeight() {
    return this.bottom - this.top;
  }

  /** @returns Position of specified corner. */
  public getCorner(corner: Corner): Point {
    switch (corner) {
      case Corner.TopLeft: {
        return this.topLeft();
      }
      case Corner.TopRight: {
        return new Point(this.right, this.top);
      }
      case Corner.BottomLeft: {
        return new Point(this.left, this.bottom);
      }
      case Corner.BottomRight: {
        return new Point(this.right, this.bottom);
      }
    }
  }

  /**
   * Inset the bounds of this rectangle.
   * @note Negative arguments will increase the size of rectangle.
   * @returns New [[Rectangle]] with modified bounds.
   */
  public inset(left: number, top: number, right: number, bottom: number) {
    return new Rectangle(this.left + left, this.top + top, this.right - right, this.bottom - bottom);
  }

  /**
   * Offsets the rectangle along the X and Y axes.
   * @returns New [[Rectangle]] with modified position.
   */
  public offset(offset: PointProps) {
    return new Rectangle(this.left + offset.x, this.top + offset.y, this.right + offset.x, this.bottom + offset.y);
  }

  /**
   * Offsets the rectangle along the X axis.
   * @returns New [[Rectangle]] with modified position along X axis.
   */
  public offsetX(offset: number) {
    return new Rectangle(this.left + offset, this.top, this.right + offset, this.bottom);
  }

  /**
   * Offsets the rectangle along the Y axis.
   * @returns New [[Rectangle]] with modified position along Y axis.
   */
  public offsetY(offset: number) {
    return new Rectangle(this.left, this.top + offset, this.right, this.bottom + offset);
  }

  /**
   * Moves the top left corner of rectangle to specified point.
   * @returns New [[Rectangle]] with modified position.
   */
  public setPosition(position: PointProps) {
    return new Rectangle(position.x, position.y, position.x + this.getWidth(), position.y + this.getHeight());
  }

  /**
   * Sets the height of the rectangle.
   * @note Only [[Edge.Bottom]] is subject to change.
   * @returns New [[Rectangle]] with modified height.
   */
  public setHeight(height: number) {
    return new Rectangle(this.left, this.top, this.right, this.top + height);
  }

  /**
   * Sets the width of the rectangle.
   * @note Only [[Edge.Right]] is subject to change.
   * @returns New [[Rectangle]] with modified width.
   */
  public setWidth(width: number) {
    return new Rectangle(this.left, this.top, this.left + width, this.bottom);
  }

  /** Checks if bounds of two rectangles match. */
  public equals(other: RectangleProps) {
    if (this.left === other.left &&
      this.top === other.top &&
      this.right === other.right &&
      this.bottom === other.bottom)
      return true;

    return false;
  }

  /**
   * Checks if point is within bounds of the rectangle.
   * @note Inclusive.
   */
  public containsPoint(point: PointProps) {
    return point.x >= this.left && point.x <= this.right && point.y >= this.top && point.y <= this.bottom;
  }

  /**
   * @returns true if this rectangle contains other rectangle.
   * @note Inclusive.
   */
  public contains(other: RectangleProps) {
    return other.left >= this.left && other.right <= this.right && other.top >= this.top && other.bottom <= this.bottom;
  }

  /** @returns New [[Rectangle]] which is contained in other rectangle. */
  public containIn(other: RectangleProps): Rectangle {
    let contained: Rectangle = this.containVerticallyIn(other);
    contained = contained.containHorizontallyIn(other);
    return contained;
  }

  /** @returns New [[Rectangle]] which is vertically contained in other rectangle. */
  public containVerticallyIn(other: RectangleProps): Rectangle {
    let contained: Rectangle = this;
    if (contained.bottom > other.bottom)
      contained = contained.offsetY(other.bottom - contained.bottom);
    if (contained.top < other.top)
      contained = contained.offsetY(other.top - contained.top);
    return contained;
  }

  /** @returns New [[Rectangle]] which is horizontally contained in other rectangle. */
  public containHorizontallyIn(other: RectangleProps): Rectangle {
    let contained: Rectangle = this;
    if (contained.right > other.right)
      contained = contained.offsetX(other.right - contained.right);
    if (contained.left < other.left)
      contained = contained.offsetX(other.left - contained.left);
    return contained;
  }

  /** @returns [[Corner.TopLeft]] position of this rectangle. */
  public topLeft(): Point {
    return new Point(this.left, this.top);
  }

  /** @returns Center point position of this rectangle. */
  public center(): Point {
    const x = this.left + (this.right - this.left) / 2;
    const y = this.top + (this.bottom - this.top) / 2;
    return new Point(x, y);
  }

  /** @returns true if this rectangle intersects other rectangle. */
  public intersects(other: RectangleProps) {
    return this.left < other.right && this.right > other.left &&
      this.top < other.bottom && this.bottom > other.top;
  }

  /**
   * Merges outer edges of this and other rectangles.
   * @returns New [[Rectangle]] with merged bounds.
   */
  public outerMergeWith(other: RectangleProps) {
    let left = this.left;
    let top = this.top;
    let right = this.right;
    let bottom = this.bottom;
    if (other.left < left)
      left = other.left;
    if (other.top < top)
      top = other.top;
    if (other.right > right)
      right = other.right;
    if (other.bottom > bottom)
      bottom = other.bottom;
    return new Rectangle(left, top, right, bottom);
  }

  /**
   * Vertically divides this rectangle into specified number of equal height segments.
   * @returns Vertical rectangle segment.
   */
  public getVerticalSegmentBounds(segmentId: number, numberOfSegments: number): Rectangle {
    const segmentHeight = this.getHeight() / numberOfSegments;

    const top = segmentId * segmentHeight;
    return this.inset(0, top, 0, 0).setHeight(segmentHeight);
  }

  /**
   * Horizontally divides this rectangle into specified number of equal width segments.
   * @returns Horizontal rectangle segment.
   */
  public getHorizontalSegmentBounds(segmentId: number, numberOfSegments: number): Rectangle {
    const segmentWidth = this.getWidth() / numberOfSegments;

    const left = segmentId * segmentWidth;
    return this.inset(left, 0, 0, 0).setWidth(segmentWidth);
  }
}
