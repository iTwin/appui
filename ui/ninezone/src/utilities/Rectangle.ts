/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Utilities */

import Point, { PointProps } from "./Point";
import Size, { SizeProps } from "./Size";

export interface RectangleProps {
  readonly left: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
}

export enum Corner {
  TopLeft,
  TopRight,
  BottomRight,
  BottomLeft,
}

export enum Edge {
  Left,
  Top,
  Right,
  Bottom,
}

export default class Rectangle implements RectangleProps {
  public static create(props: RectangleProps) {
    return new Rectangle(props.left, props.top, props.right, props.bottom);
  }

  public static createFromSize(size: SizeProps) {
    return new Rectangle(0, 0, size.width, size.height);
  }

  public constructor(public readonly left = 0, public readonly top = 0, public readonly right = 0, public readonly bottom = 0) {
  }

  public getSize() {
    const width = this.getWidth();
    const height = this.getHeight();
    return new Size(width, height);
  }

  public getWidth() {
    return this.right - this.left;
  }

  public getHeight() {
    return this.bottom - this.top;
  }

  public getCorner(corner: Corner) {
    switch (corner) {
      case Corner.TopLeft: {
        return this.topLeft();
      }
      case Corner.TopRight: {
        return new Point(this.right, this.top);
      }
      case Corner.BottomLeft: {
        return new Point(this.left, this.bottom);
      }
      case Corner.BottomRight: {
        return new Point(this.right, this.bottom);
      }
    }
  }

  public inset(left: number, top: number, right: number, bottom: number) {
    return new Rectangle(this.left + left, this.top + top, this.right - right, this.bottom - bottom);
  }

  /** Offsets the rectangle along the X and Y axes. */
  public offset(offset: PointProps) {
    return new Rectangle(this.left + offset.x, this.top + offset.y, this.right + offset.x, this.bottom + offset.y);
  }

  /** Offsets the rectangle along the X axis. */
  public offsetX(offset: number) {
    return new Rectangle(this.left + offset, this.top, this.right + offset, this.bottom);
  }

  /** Offsets the rectangle along the Y axis. */
  public offsetY(offset: number) {
    return new Rectangle(this.left, this.top + offset, this.right, this.bottom + offset);
  }

  /** Moves the top left corner of rectangle to specified point. */
  public setPosition(position: PointProps) {
    return new Rectangle(position.x, position.y, position.x + this.getWidth(), position.y + this.getHeight());
  }

  /** Sets the height of the rectangle. */
  public setHeight(height: number) {
    return new Rectangle(this.left, this.top, this.right, this.top + height);
  }

  /** Sets the width of the rectangle. */
  public setWidth(width: number) {
    return new Rectangle(this.left, this.top, this.left + width, this.bottom);
  }

  /** Checks if bounds of two rectangles match. */
  public equals(other: RectangleProps) {
    if (this.left === other.left &&
      this.top === other.top &&
      this.right === other.right &&
      this.bottom === other.bottom)
      return true;

    return false;
  }

  /** Checks if point is within bounds of the rectangle. */
  public containsPoint(point: PointProps) {
    return point.x >= this.left && point.x <= this.right && point.y >= this.top && point.y <= this.bottom;
  }

  /** Checks if rectangle is within bounds of other rectangle. */
  public containsRectangle(other: RectangleProps) {
    return other.left >= this.left && other.right <= this.right && other.top >= this.top && other.bottom <= this.bottom;
  }

  /** Contains this rectangle within other rectangle. Returns bounds of contained rectangle. */
  public containIn(other: RectangleProps): Rectangle {
    let contained: Rectangle = this.containVerticallyIn(other);
    contained = contained.containHorizontallyIn(other);
    return contained;
  }

  /** Vertically contains this rectangle within other rectangle. Returns bounds of contained rectangle. */
  public containVerticallyIn(other: RectangleProps): Rectangle {
    let contained: Rectangle = this;
    if (contained.bottom > other.bottom)
      contained = contained.offsetY(other.bottom - contained.bottom);
    if (contained.top < other.top)
      contained = contained.offsetY(other.top - contained.top);
    return contained;
  }

  /** Horizontally contains this rectangle within other rectangle. Returns bounds of contained rectangle. */
  public containHorizontallyIn(other: RectangleProps): Rectangle {
    let contained: Rectangle = this;
    if (contained.right > other.right)
      contained = contained.offsetX(other.right - contained.right);
    if (contained.left < other.left)
      contained = contained.offsetX(other.left - contained.left);
    return contained;
  }

  /** Returns top left point of the rectangle. */
  public topLeft(): Point {
    return new Point(this.left, this.top);
  }

  /** Returns center point of the rectangle */
  public center(): Point {
    const x = this.left + (this.right - this.left) / 2;
    const y = this.top + (this.bottom - this.top) / 2;
    return new Point(x, y);
  }

  /** Returns true if rectangle intersects other rectangle. */
  public intersects(other: RectangleProps) {
    return this.left < other.right && this.right > other.left &&
      this.top < other.bottom && this.bottom > other.top;
  }

  /** Merges two rectangles by the outer edges. */
  public outerMergeWith(other: RectangleProps) {
    let left = this.left;
    let top = this.top;
    let right = this.right;
    let bottom = this.bottom;
    if (other.left < left)
      left = other.left;
    if (other.top < top)
      top = other.top;
    if (other.right > right)
      right = other.right;
    if (other.bottom > bottom)
      bottom = other.bottom;
    return new Rectangle(left, top, right, bottom);
  }
}
