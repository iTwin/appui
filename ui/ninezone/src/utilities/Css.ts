/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Utilities */

import { Rectangle, RectangleProps } from "./Rectangle";
import { Point, PointProps } from "./Point";

/** CSS helpers. */
export class Css {
  /** @returns Value in pixels. */
  public static toPx(px: number): string {
    return px + "px";
  }

  /** @returns Value in percentage. */
  public static toPercentage(px: number): string {
    return px + "%";
  }
}

/** React.CSSProperties helpers. */
export class CssProperties {
  /** @returns CSS properties that describe bounds (top, left, height, width). */
  public static fromBounds(props: RectangleProps): React.CSSProperties {
    const rectangle = Rectangle.create(props);
    return {
      height: rectangle.getHeight(),
      left: rectangle.left,
      top: rectangle.top,
      width: rectangle.getWidth(),
    };
  }

  /** @returns CSS properties that describe position (top, left). */
  public static fromPosition(props: PointProps): React.CSSProperties {
    const point = Point.create(props);
    return {
      left: point.x,
      top: point.y,
    };
  }
}
