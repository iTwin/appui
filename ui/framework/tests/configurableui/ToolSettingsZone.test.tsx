/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { expect } from "chai";
import { mount } from "enzyme";

import TestUtils from "../TestUtils";
import {
  ZoneState,
  WidgetState,
  ToolSettingsZone,
  ConfigurableUiManager,
  FrontstageDefProps,
  ToolUiProvider,
  ConfigurableCreateInfo,
  FrontstageManager,
  ToolItemDef,
  ItemPropsList,
} from "../../src";

import { RectangleProps } from "@bentley/ui-ninezone/lib/utilities/Rectangle";
import ToolbarIcon from "@bentley/ui-ninezone/lib/toolbar/item/Icon";

describe("ToolSettingsZone", () => {

  const bounds: RectangleProps = {
    left: 0,
    top: 0,
    right: 100,
    bottom: 200,
  };

  class Tool2UiProvider extends ToolUiProvider {
    constructor(info: ConfigurableCreateInfo, options: any) {
      super(info, options);

      this.toolSettingsNode = <Tool2Settings />;
    }

    public execute(): void {
    }
  }

  class Tool2Settings extends React.Component {
    public render(): React.ReactNode {
      return (
        <div>
          <table>
            <tbody>
              <tr>
                <th>Type</th>
                <th>Input</th>
              </tr>
              <tr>
                <td>Month</td>
                <td> <input type="month" /> </td>
              </tr>
            </tbody>
          </table>
        </div>
      );
    }
  }

  const testToolId = "ToolSettingsZone-TestTool";

  before(async () => {
    await TestUtils.initializeUiFramework();

    const commonItemsList: ItemPropsList = {
      items: [
        {
          toolId: testToolId,
          iconClass: "icon-home",
        },
      ],
    };

    const frontstageProps: FrontstageDefProps = {
      id: "ToolSettingsZone-TestFrontstage",
      defaultToolId: "PlaceLine",
      defaultLayout: "FourQuadrants",
      contentGroup: "TestContentGroup4",
      defaultContentId: "TestContent1",

      topCenter: {
        defaultState: ZoneState.Open,
        allowsMerging: false,
        widgetProps: [
          {
            defaultState: WidgetState.Open,
            isFreeform: false,
            iconClass: "icon-home",
            labelKey: "SampleApp:Test.my-label",
            isToolSettings: true,
          },
        ],
      },
    };

    ConfigurableUiManager.loadCommonItems(commonItemsList);
    ConfigurableUiManager.registerControl(testToolId, Tool2UiProvider);
    ConfigurableUiManager.loadFrontstage(frontstageProps);
  });

  it("mount ToolSettingsZone with active Tool Settings", () => {
    const frontstageDef = FrontstageManager.findFrontstageDef("ToolSettingsZone-TestFrontstage");
    expect(frontstageDef).to.not.be.undefined;

    if (frontstageDef) {
      FrontstageManager.setActiveFrontstageDef(frontstageDef);

      const toolItemDef = ConfigurableUiManager.findItem(testToolId);
      expect(toolItemDef).to.not.be.undefined;
      expect(toolItemDef).to.be.instanceof(ToolItemDef);

      if (toolItemDef) {
        FrontstageManager.setActiveToolId(testToolId);
        expect(FrontstageManager.activeToolId).to.eq(testToolId);

        const wrapper = mount(<ToolSettingsZone bounds={bounds} />);

        const toolbarIcon = wrapper.find(ToolbarIcon);
        toolbarIcon.simulate("click");

        wrapper.unmount();
      }
    }
  });

});
