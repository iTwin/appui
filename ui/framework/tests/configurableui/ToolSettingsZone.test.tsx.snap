// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolSettingsZone renders correctly 1`] = `""`;
